os_file_is_readable()

justinmk opened the pull request:
Use `uv_fs_access()` instead of the traditional Vim open() approach

equalsraf committed:
win_defs.h: define R_OK, W_OK for MSVC.

justinmk committed:
fs.c: implement os_file_is_readable()

Use access() because:

- We already use it for os_file_is_writable()
- Vim's old check_file_readonly() ends up using access() after all.

justinmk committed:
test: cover os_file_is_readable()

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -279,13 +279,24 @@ bool os_file_is_readonly(const char *name)
+bool os_file_is_readable(const char *name)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  return (access(name, R_OK) == 0);
```
@equalsraf I used `access()` because we're already using that in `os_file_is_writable()` and I didn't see any workarounds for it in #810. Ok?

equalsraf replied:
Should be ok, just need to add a define for R_OK to 04 ([docs](https://msdn.microsoft.com/en-us/library/1w06ktdy.aspx))

justinmk replied:
From that doc, it appears I may need to check 04 (Read-only) *and* 06 (Read and write). There is also a confusing note:

> This function only checks whether the file and directory are read-only or not, it does not check the filesystem security settings. For that you need an access token. 

I guess we'll find out if it works, once the tests are running. @equalsraf The unit tests (`make unittest`) do not require the lua-client; have you had any luck running them with MSVC?

ZyX-I commented on src/nvim/os/fs.c:
```diff
@@ -279,13 +279,24 @@ bool os_file_is_readonly(const char *name)
+///
+/// @return 0 if `name` is not writable,
+/// @return 1 if `name` is writable,
+/// @return 2 for a directory which we have rights to write into.
```
You have bool return type. And are not returning 2 under any circumstances.

ZyX-I commented on src/nvim/os/fs.c:
```diff
@@ -279,13 +279,24 @@ bool os_file_is_readonly(const char *name)
 
+/// Check if a file is readable.
+///
+/// @return 0 if `name` is not writable,
```
writable

ZyX-I commented on src/nvim/os/fs.c:
```diff
@@ -279,13 +279,24 @@ bool os_file_is_readonly(const char *name)
+/// Check if a file is readable.
+///
+/// @return 0 if `name` is not writable,
+/// @return 1 if `name` is writable,
```
readable

justinmk commented:
Looking at Vim's `os_win32.c:mch_open()`, it uses `_wopen`. Since `_waccess` exists I'm guessing we actually need to use that instead of `_access`, combined with `enc_to_utf16()` as @equalsraf  mentioned elsewhere.

The good news is that libuv already provides `uv_fs_access()`, I should have checked that before :/

justinmk committed:
os_file_is_*: libuv impl

- uv_fs_access() is far more robust than access(). In particular, it
  handles utf16 paths (Windows).
- Still need R_OK, W_OK in win_defs.h
